ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala

```diff
@@ -871,6 +898,56 @@ private class PostgresLedgerDao(
         DirectExecutionContext)
   }
 
+  private val SQL_SELECT_PARTIES =
+    SQL("select party, display_name, ledger_offset, explicit from parties")
+
+  case class ParsedPartyData(
+      party: String,
+      displayName: Option[String],
+      ledgerOffset: Long,
+      explicit: Boolean)
+
+  private val PartyDataParser: RowParser[ParsedPartyData] =
+    Macro.parser[ParsedPartyData](
+      "party",
+      "display_name",
+      "ledger_offset",
+      "explicit"
+    )
+
+  override def getParties: Future[List[PartyDetails]] =
+    dbDispatcher.executeSql { implicit conn =>
+      SQL_SELECT_PARTIES
+        .as(PartyDataParser.*)
+        .map(d => PartyDetails(Party.assertFromString(d.party), d.displayName, true))
+    }
+
+  private val SQL_INSERT_PARTY =
+    SQL("""insert into parties(party, display_name, ledger_offset, explicit)
+        |select {party}, {display_name}, ledger_end, 'true'
+        |from parameters""".stripMargin)
```

Comment:
Maybe I'm mistaken, but the ledger_end is updated AFTER inserting the transaction and updating the active contracts, which means the party gets the offset of the previous entry (checkpoint, rejection, ...). Or am I missing something?

Reply:
From the table definition:
```
  -- True iff the party was added explicitly through an API call
  explicit bool not null,
  -- For implicitly added parties: the offset of the transaction that introduced the party
  -- For explicitly added parties: the ledger end at the time when the party was added
  ledger_offset bigint
```

The above SQL statement is only for explicitly inserting parties through the ledger API, which is done outside of the persistence queue used for ledger entries.
The main purpose of this field is to allow a future migration to insert "party added" ledger entries at the correct place. This will not be an easy migration, as inserting new ledger entries in the middle of the stream means transactions will get new offsets.

Reply:
oh, right. sorry I missed that. thank you for the explanation.